### PR TITLE
Deprecate DISABLE_EXCEPTION_CATCHING=2

### PR DESCRIPTION
**Pull request by sbc100**
We can instead imply this mode by the presence of the
`EXCEPTION_CATCHING_ALLOWED` list.

This turns `DISABLE_EXCEPTION_CATCHING` into a binary option
which is easier to deal with both internally and externally.

### CODE REVIEW

File: ChangeLog.md [hunk 1]
```diff
@@ -20,7 +20,12 @@ See docs/process.md for more on how version tagging works.
 
 Current Trunk
 -------------
-- values returns from `pathconf` now match the defintions found in header files
+- `DISABLE_EXCEPTION_CATCHING=2` is now deprecated since it can be inferred from
+  the presence of the `EXCEPTION_CATCHING_ALLOWED` list.  This makes
+  `DISABLE_EXCEPTION_CATCHING` a simple binary option (0 or 1) which defaults to
+  0 which will be set to 1 internally if `EXCEPTION_CATCHING_ALLOWED` list is
+  specified.
+- values returns from `pathconf` now match the definitions found in header files
```
Comment (kripken) [inline ChangeLog.md:28]:
```suggestion
- values returned from `pathconf` now match the definitions found in header files
```
(from before, I know)

File: emcc.py [hunk 2]
```diff
@@ -1578,6 +1578,14 @@ def default_setting(name, new_default):
       default_setting('ERROR_ON_UNDEFINED_SYMBOLS', 0)
       default_setting('WARN_ON_UNDEFINED_SYMBOLS', 0)
 
+    if shared.Settings.EXCEPTION_CATCHING_ALLOWED:
+      if 'DISABLE_EXCEPTION_CATCHING' in settings_key_changes:
+        if settings_key_changes['DISABLE_EXCEPTION_CATCHING'] == '2':
+          diagnostics.warning('deprecated', 'DISABLE_EXCEPTION_CATCHING is not longer needed when specifying EXCEPTION_CATCHING_ALLOWED')
```
Comment (kripken) [inline emcc.py:1584]:
```suggestion
          diagnostics.warning('deprecated', 'DISABLE_EXCEPTION_CATCHING is no longer needed when specifying EXCEPTION_CATCHING_ALLOWED')
```

File: src/settings.js [hunk 3]
```diff
@@ -640,25 +640,23 @@ var LZ4 = 0;
...
-// DISABLE_EXCEPTION_CATCHING = 2 - disable exception catching, but enables
-//                                  catching in list of allowed functions
+//
 // XXX note that this removes *catching* of exceptions, which is the main
 //     issue for speed, but you should build source files with
 //     -fno-exceptions to really get rid of all exceptions code overhead,
 //     as it may contain thrown exceptions that are never caught (e.g.
 //     just using std::vector can have that). -fno-rtti may help as well.
 //
+// This option is ignoed if EXCEPTION_CATCHING_ALLOWED is specified.
```
Comment (kripken) [inline src/settings.js:653]:
```suggestion
// This option is ignored if EXCEPTION_CATCHING_ALLOWED is specified.
```

File: src/settings.js [hunk 4]
```diff
@@ -640,25 +640,23 @@ var LZ4 = 0;
...
+//
 // [compile+link] - affects user code at compile and system libraries at link
 var DISABLE_EXCEPTION_CATCHING = 1;
 
-// Enables catching exception in the listed functions only, if
-// DISABLE_EXCEPTION_CATCHING = 2 is set
+// Enables catching exception but only in the listed functions.  This
+// option overrides DISABLE_EXCEPTION_CATCHING.
```
Comment (kripken) [inline src/settings.js:659]: what does "override" mean? emcc.py has error handling to error on conflicting flags, IIUC, so neither overrides the other?

Comment (sbc100) [reply src/settings.js]: You are correct, I re-worded.

File: ChangeLog.md [hunk 5]
```diff
@@ -20,7 +20,12 @@ See docs/process.md for more on how version tagging works.
 
 Current Trunk
 -------------
-- values returns from `pathconf` now match the defintions found in header files
+- `DISABLE_EXCEPTION_CATCHING=2` is now deprecated since it can be inferred from
+  the presence of the `EXCEPTION_CATCHING_ALLOWED` list.  This makes
+  `DISABLE_EXCEPTION_CATCHING` a simple binary option (0 or 1) which defaults to
+  0 which will be set to 1 internally if `EXCEPTION_CATCHING_ALLOWED` list is
+  specified.
+- Values returned from `pathconf` now match the definitions found in header files
```
Comment (aheejin) [inline ChangeLog.md:28]: Is this supposed to be in another PR?

File: emcc.py [hunk 6]
```diff
@@ -1578,6 +1578,14 @@ def default_setting(name, new_default):
       default_setting('ERROR_ON_UNDEFINED_SYMBOLS', 0)
       default_setting('WARN_ON_UNDEFINED_SYMBOLS', 0)
 
+    if shared.Settings.EXCEPTION_CATCHING_ALLOWED:
+      if 'DISABLE_EXCEPTION_CATCHING' in settings_key_changes:
```
Comment (aheejin) [inline emcc.py:1582]: How is this line different from
```py
if shared.Settings.DISABLE_EXCEPTION_CATCHING:
```
? If they are different, should we also check if both `shared.Settings.DISABLE_EXCEPTION_CATCHING` and `shared.Settings.EXCEPTION_CATCHING_ALLOWE` are set?

File: emcc.py [hunk 7]
```diff
@@ -1968,7 +1976,7 @@ def check_memory_setting(setting):
        sanitize:
       shared.Settings.EXPORTED_FUNCTIONS += ['_malloc', '_free']
 
-    if shared.Settings.DISABLE_EXCEPTION_CATCHING != 1:
```
Comment (aheejin) [inline emcc.py:1971]: So this means the code below didn't run when `DISABLE_EXCEPTION_CATCHING == 2` before. I wonder how it was correct..? The same for the similar line above.

File: src/parseTools.js [hunk 8]
```diff
@@ -873,8 +873,8 @@ function makeStructuralReturn(values) {
 }
 
 function makeThrow(what) {
-  if (ASSERTIONS && DISABLE_EXCEPTION_CATCHING == 1) {
-    what += ' + " - Exception catching is disabled, this exception cannot be caught. Compile with -s DISABLE_EXCEPTION_CATCHING=0 or DISABLE_EXCEPTION_CATCHING=2 to catch."';
+  if (ASSERTIONS && DISABLE_EXCEPTION_CATCHING) {
+    what += ' + " - Exception catching is disabled, this exception cannot be caught. Compile with -s NO_DISABLE_EXCEPTION_CATCHING or -s EXCEPTION_CATCHING_ALLOWED=[..] to catch."';
```
Comment (aheejin) [inline src/parseTools.js:877]: Do we have a setting called `NO_DISABLE_EXCEPTION_CATCHING` somewhere?

File: src/settings.js [hunk 9]
```diff
@@ -640,25 +640,26 @@ var LZ4 = 0;
 
 // Disables generating code to actually catch exceptions. This disabling is on
 // by default as the overhead of exceptions is quite high in size and speed
-// currently (in the future, wasm should improve that). When exceptions are
+// currently (in the future, wasm should improve that).  When exceptions are
```
Comment (aheejin) [inline src/settings.js:643]:
```suggestion
// currently (in the future, wasm should improve that). When exceptions are
```
I know this is a matter of preference, but this paragraph uses single spaces so I think it'd be better to be consistent within the paragraph.

File: tests/test_core.py [hunk 10]
```diff
@@ -1236,15 +1236,16 @@ def test_exceptions_3(self):
     self.do_run('src.js', 'Caught exception: Hello\nDone.', args=['2'], no_build=True)
 
   def test_exceptions_allowed(self):
-    self.set_setting('DISABLE_EXCEPTION_CATCHING', 2)
```
Comment (aheejin) [inline tests/test_core.py:1239]: Shouldn't we set it to 0 instead and not delete the line? The same for other functions.

File: tests/test_core.py [hunk 11]
```diff
@@ -1236,15 +1236,16 @@ def test_exceptions_3(self):
...
     # mismatches are fixed in fixImports() function in JS glue code.
     self.set_setting('EXCEPTION_CATCHING_ALLOWED', ["_Z12somefunctionv"])
     # otherwise it is inlined and not identified
     self.set_setting('INLINING_LIMIT', 50)
 
     self.do_core_test('test_exceptions_allowed.cpp')
-    size = len(open('test_exceptions_allowed.js').read())
+    size = os.path.getsize('test_exceptions_allowed.js')
+    if self.is_wasm():
+      size += os.path.getsize('test_exceptions_allowed.wasm')
```
Comment (aheejin) [inline tests/test_core.py:1248]: This PR seems to contain the contents of #13734, here and several elsewhere in this file.

File: tools/building.py [hunk 12]
```diff
@@ -483,9 +483,9 @@ def llvm_backend_args():
   args = ['-combiner-global-alias-analysis=false']
 
   # asm.js-style exception handling
-  if Settings.DISABLE_EXCEPTION_CATCHING != 1:
+  if not Settings.DISABLE_EXCEPTION_CATCHING:
     args += ['-enable-emscripten-cxx-exceptions']
-  if Settings.DISABLE_EXCEPTION_CATCHING == 2:
+  if Settings.EXCEPTION_CATCHING_ALLOWED:
     allowed = ','.join(Settings.EXCEPTION_CATCHING_ALLOWED or ['__fake'])
```
Comment (aheejin) [inline tools/building.py:489]: Do we still need `__fake`? Not sure why it was there in the first place, but now we don't enter this `if` if `EXCEPTION_CATCHING_ALLOWED`is empty anyway.

File: ChangeLog.md (hunk 5 above)
Comment (sbc100) [reply ChangeLog.md:28]: I just fixed a couple of typos in that line here while I was adding the new entry.

File: emcc.py (hunk 6 above)
Comment (sbc100) [reply emcc.py:1582]: This is set if the user specified it explicitly on the command line (regardless of if the value matches the default).  

Doing that here allows me to show a warning (or error) if the user ever passed both these options at the same time.

File: emcc.py (hunk 7 above)
Comment (sbc100) [reply emcc.py:1971]: No the code below did run in that case.    If `DISABLE_EXCEPTION_CATCHING` was 0 or 2 it would add the exported functions.

Because we set `DISABLE_EXCEPTION_CATCHING=0` when the `EXCEPTION_CATCHING_ALLOWED` list is specified (above) we don't need to explicitly check for `EXCEPTION_CATCHING_ALLOWED` here.

File: src/parseTools.js (hunk 8 above)
Comment (sbc100) [reply src/parseTools.js:877]: All settings can be prefixed with `NO_` to set them to zero as an alternative to `=0`.   

e.g. `-s NO_EXIT_RUNTIME` is the same as `-s EXIT_RUNTIME=0`

Personally I think this option reads better that way but I can revert this part if you prefer.

File: src/settings.js (hunk 9 above)
Comment (sbc100) [reply src/settings.js:643]: OK.

File: tests/test_core.py (hunk 10 above)
Comment (sbc100) [reply tests/test_core.py:1239]: No because we set `EXCEPTION_CATCHING_ALLOWED` below and setting them both at the same time is a warning (and we are building with `-Werror`)

File: tests/test_core.py (hunk 11 above)
Comment (sbc100) [reply tests/test_core.py:1248]: Yes, there are stacked but both targetting main.   Now that #13734,has landed the diff should look cleaner..

File: tools/building.py (hunk 12 above)
Comment (sbc100) [reply tools/building.py:489]: Hmm indeed, I think this can just be removed.

File: emcc.py (hunk 6 above)
Comment (aheejin) [reply emcc.py:1582]: Then shouldn't we check if both of those two flags are specified in
1. One in `Settings` and the other in the command line (and vice versa)
2. Both in `Settings`
3. Both in the command line

Maybe I could be mistaken.

File: ChangeLog.md (hunk 5 above)
Comment (aheejin) [reply ChangeLog.md:28]: Right it was an existing line. I missed it, sorry.

File: src/parseTools.js (hunk 8 above)
Comment (aheejin) [reply src/parseTools.js:877]: No I didn't know there is always a matching `NO_` option for every setting. Thanks for letting me know.

File: tests/test_core.py (hunk 10 above)
Comment (aheejin) [reply tests/test_core.py:1239]: No I meant 0, not 1. The default setting for `DISABLE_EXCEPTION_CATCHING` is 1, so I thought we needed to turn it off to catch exceptions. Settings `DISABLE_EXCEPTION_CATCHING == 0` and `EXCEPTION_CATCHING_ALLOWED` at the same time is an error? 

Or does setting `EXCEPTION_CATCHING_ALLOWED` automatically imply `DISABLE_EXCEPTION_CATCHING == 0` or something?

File: tests/test_core.py [hunk 13]
```diff
@@ -1278,14 +1277,14 @@ def test_exceptions_allowed(self):
...
     print('fake_size: %d' % fake_size)
     print('disabled_size: %d' % disabled_size)
-    self.assertEqual(empty_size, fake_size)
+    # empty list acts the same a full disable
+    self.assertEqual(empty_size, disabled_size)
     # big change when we disable exception catching of the function
     self.assertGreater(size - empty_size, 0.01 * size)
     # full disable can remove a little bit more
-    self.assertLess(disabled_size, empty_size)
+    self.assertLess(disabled_size, fake_size)
```
Comment (aheejin) [inline tests/test_core.py:1285]: It still seems to contain some of the previous PR?

Comment (sbc100) [reply tests/test_core.py:1285]: These changes are actually part of this one now.

File: tests/test_core.py (hunk 10 above)
Comment (sbc100) [reply tests/test_core.py:1239]: Yes exactly, setting EXCEPTION_CATCHING_ALLOWED implied DISABLE_EXCEPTION_CATCHING.

Also setting `DISABLE_EXCEPTION_CATCHING` to anything, including 0, at the same time as `EXCEPTION_CATCHING_ALLOWED` an error or warning (=2 is a warning =1, or =0 is an error).  They should never appear together .. although we need to continue to support =2 for backward compat (we give a warning in this case).

File: emcc.py (hunk 6 above)
Comment (kripken) [reply emcc.py:1582]: In theory yes, but I don't think we are careful about local modifications to `settings.js`. That is, users should use the commandline. Changes to `settings.js` by an emscripten developer would assume they know the risk of changing internals.

**Review by kripken — APPROVED** [review]

File: emcc.py (hunk 6 above)
Comment (aheejin) [reply emcc.py:1582]: If users are supposed to use the command line, then we should check the case when both of them are in `settings_key_change` at least?

Also most of these checking lines in this function directly refer to not `settings_key_change` but `shared.Settings` directly, which is weird if users are supposed use the command line.

File: tests/test_core.py (hunk 10 above)
Comment (aheejin) [reply tests/test_core.py:1239]: I see. But why is setting `DISABLE_EXCEPTION_CATCHING` to 0 also an error? Are gonna do that internally anyway?

File: emcc.py (hunk 6 above)
Comment (sbc100) [reply emcc.py:1592]: In general yes, although I think this code covers the cases we care about.

I don't think (2) is much of an issue.  I don't think folks should be directly modifying `settings.js`, at least not without also also modifying code like this that is somewhat aware of the default.   So I think its fair to say that if `EXCEPTION_CATCHING_ALLOWED` is set then its the user that set it.

That means if these two conditions are both two they specified both on the command line.  That is exactly the case we care about erroring/warning for.   I agree we could possibly be more robust here and not make assumptions about the default values.

I added a comment which should hopefully make it a little more clear what is going on.

**Comment by sbc100** [comment]
Added a test for the new error/warning

File: emcc.py (hunk 6 above)
Comment (aheejin) [reply emcc.py:1592]: Sorry, what I was asking is, if we only care about the command line,
why are we not checking `settings_key_change` for `EXCEPTION_CATCHING_ALLOWED` as we do for `DISABLE_EXCEPTION_CATCHING`? We are checking `settings_key_changes` for one and checking `shared.Settings` for the other. This is the part I don't understand.

Also, if we only care about command lines, why are all other lines in this function directly checking `shared.Settings` and now `settings_key_changes`?

File: tests/test_core.py (hunk 10 above)
Comment (sbc100) [reply tests/test_core.py:1239]: It seems cleaner to me that they should never be specified together.  Don't you think they should be 100% mutually exclusive?    

That fact that internally we set `DISABLE_EXCEPTION_CATCHING` to zero when `EXCEPTION_CATCHING_ALLOWED` is specified is kind of an implementation detail I think.

File: tests/test_core.py [hunk 14]
```diff
@@ -1278,33 +1277,49 @@ def test_exceptions_allowed(self):
     print('empty_size: %d' % empty_size)
     print('fake_size: %d' % fake_size)
     print('disabled_size: %d' % disabled_size)
-    self.assertEqual(empty_size, fake_size)
+    # empty list acts the same a full disable
```
Comment (aheejin) [inline tests/test_core.py:1280]:
```suggestion
    # empty list acts the same as a full disable
```

File: tests/test_core.py (hunk 13 above)
Comment (aheejin) [reply tests/test_core.py:1285]: If empty list act as a full disable and `empty_size` is equal to `fake_size` as you wrote, what is `disabled_size` here? It also says this is full disable... Which one is real full disable? `empty_size` of `disabled_size`? Does setting `EXCEPTION_CATCHING_ALLOWED` to a nonempty list have any other effect?

File: emcc.py (hunk 6 above)
Comment (sbc100) [reply emcc.py:1592]: > If users are supposed to use the command line, then we should check the case when both of them are in `settings_key_change` at least?
> 
> Also most of these checking lines in this function directly refer to not `settings_key_change` but `shared.Settings` directly, which is weird if users are supposed use the command line.

Technically you are correct yes, we should also check for `EXCEPTION_CATCHING_ALLOW` in `settings_key_changes`.

Maybe this latest version is more explicit.

File: tests/test_core.py (hunk 10 above)
Comment (aheejin) [reply tests/test_core.py:1239]: True, but `DISABLE_EXCEPTION_CATCHING` is still an exposed settings variable, so I think a warning can be fine, but an error seems weird, especially specifying both of them is semantically correct. Also this will not be backward compatible.

File: emcc.py (hunk 6 above)
Comment (aheejin) [reply emcc.py]: Thanks, but the problem here is my lack of knowledge of the code around here ;( Sorry. Can someone answer why we check `shared.Settings` directly in the code within this function, if we expect the users to use only the command line? I asked why we were not using `settings_key_changes` for both settings, but other code around this line all check `shared.Settings` directly, so I'm wondering why this should be, or if this should be, different, and how all this works.

When are we merging `settings_key_changes` to `shared.Settings`? Before running this code or after? If after, checking `settings.Changes` only makes sense, but if so you wouldn't have checked `settings_key_changes` in the first place...

Sorry for many questions. If taking offline is easier it's fine too.

File: tests/test_core.py (hunk 13 above)
Comment (sbc100) [reply tests/test_core.py:1285]: disabled_size is when no options are given and exceptions are therefore completely disabled.

With this new change settings `EXCEPTION_CATCHING_ALLOWED` to empty list is the same as completely disabled.

Settings `EXCEPTION_CATCHING_ALLOWED` to a non-empty list internally sets `DISABLE_EXCEPTION_CATCHING=0` so all the JS-side assumes exceptions can be caught.  The only difference is in the underlying call to clang which either limits (or does not limit) the functions which can catch (IIUC).

File: emcc.py (hunk 6 above)
Comment (sbc100) [reply emcc.py]: `shared.Settings` reflects that actual current settings which is the sum of:

a) the defaults
b) any user command line settings
c) any changes made by the running of the code here (which updates a lot of settings).

In almost all cases `shared.Settings` is the only thing we care about because it represent the actual current setup.

`settings_key_changes` is for cases were we actually care about what the user specified on the commandline as opposed to (a) or (c).    Its mostly used for diagnostics stuff like this.    For example in this case we need to use it so we can tell the different between the default `=0` and the command line `=0`.

I will make a couple of PRs to clean this up a bit and give `settings_key_changes` a better name :)

**Comment by sbc100** [comment]
`fake_size` is not the same as `disabled_size` (or `empty_size`), its slightly bigger because the compiler driver thinks there is at least one function which can catch, so it has to enable exceptions on the JS side (it sets `DISABLE_EXCEPTION_CATCHING=0`).     

`disabled_size` (or `empty_size`) are different because the compiler does not set `DISABLE_EXCEPTION_CATCHING=0`.

File: emcc.py [hunk 15]
```diff
@@ -1588,6 +1588,18 @@ def default_setting(name, new_default):
       default_setting('ERROR_ON_UNDEFINED_SYMBOLS', 0)
       default_setting('WARN_ON_UNDEFINED_SYMBOLS', 0)
 
+    if 'DISABLE_EXCEPTION_CATCHING' in settings_key_changes and 'EXCEPTION_CATCHING_ALLOWED' in settings_key_changes:
+      # If we get here then the user specified both DISABLE_EXCEPTION_CATCHING and EXCEPTION_CATCHING_ALLOWED
+      # on the command line.  This is not longer valid so report either an error or a warning (for
```
Comment (aheejin) [inline emcc.py:1593]:
```suggestion
      # on the command line.  This is no longer valid so report either an error or a warning (for
```

File: tests/test_core.py [hunk 16]
```diff
@@ -1278,33 +1277,54 @@ def test_exceptions_allowed(self):
...
 
+  def test_exceptions_allowed_misuse(self):
+    self.set_setting('EXCEPTION_CATCHING_ALLOWED', ['foo'])
+
+    # Test old =2 setting for DISABLE_EXCEPTION_CATCHING
+    self.set_setting('DISABLE_EXCEPTION_CATCHING', 2)
+    err = self.expect_fail([EMCC, test_file('hello_world.c')] + self.get_emcc_args())
+    self.assertContained('error: DISABLE_EXCEPTION_CATCHING is no longer needed when specifying EXCEPTION_CATCHING_ALLOWED [-Wdeprecated] [-Werror]', err)
+
+    # =0 and =2 should be a hard error
```
Comment (aheejin) [inline tests/test_core.py:1314]:
```suggestion
    # =0 and =1 should be a hard error
```
I think having 0 as a warning is more intuitive though.

**Review by aheejin — COMMENTED** [review]
> `shared.Settings` reflects that actual current settings which is the sum of:
>
> a) the defaults
> b) any user command line settings
> c) any changes made by the running of the code here (which updates a lot of settings).
>
> In almost all cases `shared.Settings` is the only thing we care about because it represent the actual current setup.
>
> `settings_key_changes` is for cases were we actually care about what the user specified on the commandline as opposed to (a) or (c).    Its mostly used for diagnostics stuff like this.    For example in this case we need to use it so we can tell the different between the default `=0` and the command line `=0`.
>
> I will make a couple of PRs to clean this up a bit and give `settings_key_changes` a better name :)

Thanks for the explanation. But sorry the same questions ;( :
- @kripken said we mostly only care about command lines. Then why are other checks within this function only check `shared.Settings` and not `settings_key_changes`? Why should this one be different?
- You said `shared.Settings` is the only thing we care about. I think that means by this point `settings_key_changes` has been merged into `shared.Settings`, right? Then why can't we check `shared.Settings` as the final merged info?

I'm commenting this separately because somehow I can't comment anymore in the original thread inline.

**Comment by sbc100** [comment]
> > `shared.Settings` reflects that actual current settings which is the sum of:
> > a) the defaults
> > b) any user command line settings
> > c) any changes made by the running of the code here (which updates a lot of settings).
> > In almost all cases `shared.Settings` is the only thing we care about because it represent the actual current setup.
> > `settings_key_changes` is for cases were we actually care about what the user specified on the commandline as opposed to (a) or (c).    Its mostly used for diagnostics stuff like this.    For example in this case we need to use it so we can tell the different between the default `=0` and the command line `=0`.
> > I will make a couple of PRs to clean this up a bit and give `settings_key_changes` a better name :)
> 
> Thanks for the explanation. But sorry the same questions ;( :
> 
> * @kripken said we mostly only care about command lines. Then why are other checks within this function only check `shared.Settings` and not `settings_key_changes`? Why should this one be different?

I think what he means is "When we want to see if a user has changed a given setting, we only need to consider about the command line".     When we want to check the setting in general (regardless is seeing it was changed or not), we tend to use `shared.Settings`.   99.9% of settings checking is done using `shared.Settings`.  We only use `settings_key_changes` when we want to be able to tell if something was set on the command line (as opposed to set in the config file).  Normally we don't care about the difference, but in some cases like this one we do.

> * You said `shared.Settings` is the only thing we care about. I think that means by this point `settings_key_changes` has been merged into `shared.Settings`, right? Then why can't we check `shared.Settings` as the final merged info?

I think I already answered this above: We only use `settings_key_changes` when we want to be able to tell if something was set on the command line (as opposed to set in the config file).  Normally we don't care about the difference.


> 
> I'm commenting this separately because somehow I can't comment anymore in the original thread inline.

File: tests/test_core.py (hunk 16 above)
Comment (sbc100) [reply tests/test_core.py:1314]: Why?   The definition of =0 today (before this change) means "all functions" .. so specifying additionally an allow list makes not sense today.. so there should be no such users.

Comment (sbc100) [reply tests/test_core.py:1314]: As of today the only value that makes sense along with the `allow list` is `=2`, right?  Why should I start allowing `=0` + `allow list` in this change?   The point of this change is to try to make these two settings as mutually exclusive as possible without breaking backwards compat.

Comment (sbc100) [reply tests/test_core.py]: OK, I made =0 into a warning too.  The difference I don't think is very important.

Comment (aheejin) [reply tests/test_core.py]: I think my interpretation of the switches were different. I thought `DISABLE_EXCEPTION_CATCHING` as a on-off switch, which we turn on whenever (in all functions or in only some of them) we want to catch exceptions, and `EXCEPTION_CATCHING_ALLOWED` as a fine-grained knob to say which function (if not specified, all functions).

I thought this interpretation was more intuitive, but I don't have strong opinions, so please revert if you feel like it.

**Comment by aheejin** [comment]
> We only use `settings_key_changes` when we want to be able to tell if something was set on the command line (as opposed to set in the config file). Normally we don't care about the difference, but in some cases like this one we do.

Can you explain a little more on what the difference between this case and other 99.9% of cases? This is mostly to understand the code myself and I don't want to block this ;) Thanks.

**Review by aheejin — APPROVED** [review]

**Comment by sbc100** [comment]
> > We only use `settings_key_changes` when we want to be able to tell if something was set on the command line (as opposed to set in the config file). Normally we don't care about the difference, but in some cases like this one we do.
> 
> Can you explain a little more on what the difference between this case and other 99.9% of cases? This is mostly to understand the code myself and I don't want to block this ;) Thanks for the long explanation!

Most of the time code only cares about the state of a given setting, so it quite rightly checks the `Settings` object

```
$ git grep Settings "*.py" | wc -l
799
```

There is also all the settings checks in JS code which I cannot easily count.  (e.g. `#if USE_PTHREADS` etc).

Occasionally we actually care about *how* a settings came to be.  i.e. did the user explicitly pass on the command line.  In those few cases (which only exist in one place) we can use `settings_key_changes`,   There are only nine uses of settings_key_change:

```
emcc.py:    strict_cmdline = settings_key_changes.get('STRICT')
emcc.py:    if settings_key_changes.get('WARN_ON_UNDEFINED_SYMBOLS') == '0':
emcc.py:    if shared.Settings.MINIMAL_RUNTIME or settings_key_changes.get('MINIMAL_RUNTIME') in ('1', '2'):
emcc.py:      if 'EXPORTED_FUNCTIONS' in settings_key_changes:
emcc.py:      if 'EXIT_RUNTIME' in settings_key_changes:
emcc.py:      if name not in settings_key_changes:
emcc.py:      if 'MODULARIZE' in settings_key_changes:
emcc.py:    if 'MEM_INIT_METHOD' in settings_key_changes:
emcc.py:      if settings_key_changes.get('WASM_BIGINT') == '0':
```

Most of these are cases where we want to know if the user explicitly enabled something.   For example:

```
     if 'EXIT_RUNTIME' in settings_key_changes:                                                                      
        exit_with_error('Explictly setting EXIT_RUNTIME not compatible with STANDALONE_WASM.  EXIT_RUNTIME will always be True for programs (with a main function) and False for reactors (not main function).')
```

Here we have a setting called `EXIT_RUNTIME` that can be set on the command but is also set implicitly in `STANDALONE_WASM`.   In this mode its not permitted to explicitly set it and the only way we can detect this is via `settings_key_changes`.   `STANDALONE_WASM` is also BTW implicitly set in some cases.

**Comment by sbc100** [comment]
Note that in the `if 'EXIT_RUNTIME' in settings_key_changes:` example I gave we actually don't care what value the setting has at all.. we only care if it was set by the user explictly.
